src/calculate_fusion.c

```diff
@@ -0,0 +1,353 @@
+#include <stdio.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <gsl/gsl_math.h>
+#include <gsl/gsl_eigen.h>
+#include "calculate_fusion.h"
+
+struct support_degree_matrix *calculate_support_degree_matrix(void)
+{
+    // if (node == NULL)
+    // {
+    //     printf("%s: Incorrect Input\n", __func__);
+    //     return;
+    // }
+    // int i = 0;
+    // while (node != NULL)
+    // {
+    //     printf("%c\n", node->sensor_name);
+    //     node = node->next;
+    //     i++;
+    //     printf("%d\n", i);
+    // }
+    float values[] = {10, 0};
+    struct support_degree_matrix *spd;
+    spd = (struct support_degree_matrix *)malloc(sizeof(struct support_degree_matrix));
+    int length = sizeof(values) / sizeof(values[0]);
+    spd->no_of_sensors = length;
+    double **arrptr = (double **)malloc(length * sizeof(double *));
+    int count = 0;
+    for (int i = 0; i < length; i++)
+    {
+        arrptr[i] = (double *)malloc(length * sizeof(double));
+    }
+    spd->sd_matrix = (double *)malloc(sizeof(double) * length * length);
+    if (arrptr == NULL || spd == NULL || spd->sd_matrix == NULL)
+    {
+        printf("%s: Unable to allocate memory!\n", __func__);
+        return NULL;
+    }
+
+    for (int i = 0; i < length; i++)
+    {
+        for (int j = 0; j < length; j++)
+        {
+            arrptr[i][j] = exp(-1 * fabs(values[i] - values[j]));
+            spd->sd_matrix[count] = arrptr[i][j];
+            count++;
+        }
+    }
+    printf("Support Degree Matrix\n");
+    for (int i = 0; i < length * length; i++)
+    {
+        printf("%f\n", spd->sd_matrix[i]);
+    }
+    free(arrptr);
```

Comment:
this free is actually just freeing `double **arrptr = (double **)malloc(length * sizeof(double *));` ,that is pointing to the 2d matrix, not the entire chunk of memory that was allocated for the matrix. If you would like to free the chunk of memory you will have to loop and free ` arrptr[i] = (double *)malloc(length * sizeof(double));` for each allocation. Maybe after you are done with spd. you can clean up the memory.